engines/sci/graphics/video32.cpp

```diff
@@ -0,0 +1,364 @@
+	_plane(nullptr),
+	_screenItem(nullptr),
+	_stretchVertical(false),
+	_priority(0),
+	_blackoutPlane(nullptr),
+	_yieldInterval(0) {}
+
+VMDPlayer::~VMDPlayer() {
+	close();
+	delete _decoder;
+}
+
+VMDPlayer::IOStatus VMDPlayer::open(const Common::String &fileName, const OpenFlags flags) {
+	if (_isOpen) {
+		error("Attempted to play %s, but another VMD was loaded", fileName.c_str());
+	}
+
+	if (_decoder->loadFile(fileName)) {
+		if (flags & kOpenFlagMute) {
+			_decoder->setVolume(0);
+		}
+		_isOpen = true;
+		return kIOSuccess;
+	} else {
+		return kIOError;
+	}
+}
+
+void VMDPlayer::init(const int16 x, const int16 y, const PlayFlags flags, const int16 boostPercent, const int16 boostStartColor, const int16 boostEndColor) {
+	_x = getSciVersion() >= SCI_VERSION_3 ? x : (x & ~1);
+	_y = y;
+	_leaveScreenBlack = flags & kPlayFlagLeaveScreenBlack;
+	_leaveLastFrame = flags & kPlayFlagLeaveLastFrame;
+	_doublePixels = flags & kPlayFlagDoublePixels;
+	_blackLines = flags & kPlayFlagBlackLines;
+	_boostPercent = 100 + (flags & kPlayFlagBoost ? boostPercent : 0);
+	_blackPalette = flags & kPlayFlagBlackPalette;
+	_stretchVertical = flags & kPlayFlagStretchVertical;
+	_boostStartColor = CLIP<int16>(boostStartColor, 0, 255);
+	_boostEndColor = CLIP<int16>(boostEndColor, 0, 255);
+}
+
+void VMDPlayer::restrictPalette(const uint8 startColor, const uint8 endColor) {
+	_startColor = startColor;
+	_endColor = endColor;
+}
+
+VMDPlayer::EventFlags VMDPlayer::kernelPlayUntilEvent(const EventFlags flags, const int16 lastFrameNo, const int16 yieldInterval) {
+	assert(lastFrameNo >= -1);
+
+	const int32 maxFrameNo = (int32)(_decoder->getFrameCount() - 1);
+
+	if ((flags & kEventFlagToFrame) && lastFrameNo > 0) {
+		_decoder->setEndFrame(MIN((int32)lastFrameNo, maxFrameNo));
+	} else {
+		_decoder->setEndFrame(maxFrameNo);
+	}
+
+	if (flags & kEventFlagYieldToVM) {
+		_yieldInterval = 3;
+		if (yieldInterval == -1 && !(flags & kEventFlagToFrame)) {
+			_yieldInterval = lastFrameNo;
+		} else if (yieldInterval != -1) {
+			_yieldInterval = MIN((int32)yieldInterval, maxFrameNo);
+		}
+	} else {
+		_yieldInterval = maxFrameNo;
+	}
+
+	return playUntilEvent(flags);
+}
+
+VMDPlayer::EventFlags VMDPlayer::playUntilEvent(const EventFlags flags) {
+	// Flushing all the keyboard and mouse events out of the event manager to
+	// avoid letting any events queued from before the video started from
+	// accidentally activating an event callback
+	for (;;) {
+		const SciEvent event = _eventMan->getSciEvent(SCI_EVENT_KEYBOARD | SCI_EVENT_MOUSE_PRESS | SCI_EVENT_MOUSE_RELEASE | SCI_EVENT_QUIT);
+		if (event.type == SCI_EVENT_NONE) {
+			break;
+		} else if (event.type == SCI_EVENT_QUIT) {
+			return kEventFlagEnd;
+		}
+	}
+
+	_decoder->pauseVideo(false);
+
+	if (flags & kEventFlagReverse) {
+		// NOTE: This flag may not work properly since SCI does not care
+		// if a video has audio, but the VMD decoder does, for some reason.
+		assert(_decoder->setReverse(true));
```

Comment:
I can attest from personal experience that it's bad form to put a call that has to happen inside an assert statement, since when the codebase is built in release mode, the entire line will be removed. You should instead get the result from the function, and assert the value on a separate line.